### PR TITLE
Add Cardano/Shelley address support, remove Cardano/Byron

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Current Cardano version is Byron (V2).  Next version is Shelley (V3), in testnet currently.
Byron support was added (incl. signing, PR #797 ), but is being removed now.
Shelley support is being added -- address part. 

## Testing instructions

Tests.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.

### CODE REVIEW

File: coins.json [hunk 1]
```diff
@@ -1290,20 +1290,20 @@
         "symbol": "ADA",
         "decimals": 6,
         "blockchain": "Cardano",
-        "derivationPath": "m/44'/1815'/0'/0/0",
-        "curve": "ed25519Extended",
-        "publicKeyType": "ed25519Extended",
+        "derivationPath": "m/1852'/1815'/0'/0/0",
```
Comment (hewigovens) [inline coins.json]: can you remove trailing space? maybe you have to change your editor setting

File: include/TrustWalletCore/TWCoinType.h [hunk 2]
```diff
@@ -29,7 +29,7 @@ enum TWCoinType {
     TWCoinTypeBitcoinCash = 145,
     TWCoinTypeBravoCoin = 282,
     TWCoinTypeCallisto = 820,
-    TWCoinTypeCardano = 1815,
+    TWCoinTypeCardano = 1815, // Note: Cardano Shelley testnet uses purpose 1852 (not 44) 1852/1815
```
Comment (hewigovens) [inline include/TrustWalletCore/TWCoinType.h]: how about mainnet? still 1852?

File: src/Cardano/AddressV3.cpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        Kind kind;
+        Data key1;
+        Data key2;
+        if (parseAndCheckV3(addr, discrimination, kind, key1, key2)) {
+            return true;
+        }
+    } catch (exception& ex) {
+    }
+    // not V3, try older
+    return TW::Cardano::AddressV2::isValid(addr);
```
Comment (hewigovens) [inline src/Cardano/AddressV3.cpp]:
```suggestion
    return AddressV2::isValid(addr);
```

File: src/Cardano/AddressV3.cpp [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    }
+}
+
+bool AddressV3::isValid(const std::string& addr) {
+    try {
+        Discrimination discrimination;
+        Kind kind;
+        Data key1;
+        Data key2;
+        if (parseAndCheckV3(addr, discrimination, kind, key1, key2)) {
```
Comment (hewigovens) [inline src/Cardano/AddressV3.cpp]: it won't throw exception?

File: src/Cardano/AddressV3.cpp [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+#include "../HexCoding.h"
+#include "../Hash.h"
+
+#include <array>
+
+using namespace TW;
+using namespace TW::Cardano;
+using namespace std;
+
+bool AddressV3::parseAndCheckV3(const std::string& addr, Discrimination& discrimination, Kind& kind, Data& key1, Data& key2) {
```
Comment (hewigovens) [inline src/Cardano/AddressV3.cpp]: maybe we can return tuple? too many inout parameters

File: src/Cardano/AddressV3.cpp [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    legacyAddressV2(other.legacyAddressV2 == nullptr ? nullptr : new AddressV2(*other.legacyAddressV2))
+{}
+
+string AddressV3::string() const {
+    std::string hrp;
+    switch (kind) {
+        case Kind_Single:
+        case Kind_Group:
+        case Kind_Account:
+            hrp = "addr"; break;
```
Comment (hewigovens) [inline src/Cardano/AddressV3.cpp]: let's add `hrp` to coins.json?

File: coins.json (hunk 1 above)
Comment (catenocrypt) [reply coins.json]: done

File: include/TrustWalletCore/TWCoinType.h (hunk 2 above)
Comment (catenocrypt) [reply include/TrustWalletCore/TWCoinType.h]: No idea, I will have to check.

File: src/Cardano/AddressV3.cpp (hunk 3 above)
Comment (catenocrypt) [reply src/Cardano/AddressV3.cpp]: done

File: src/Cardano/AddressV3.cpp (hunk 4 above)
Comment (catenocrypt) [reply src/Cardano/AddressV3.cpp]: Bech method, invoked by parseAndCheckV3, may throw if format is not correct.  I have moved the catch one method higher.

File: src/Cardano/AddressV3.cpp (hunk 6 above)
Comment (catenocrypt) [reply src/Cardano/AddressV3.cpp]: Yes, done.  The switch is still needed, though.

File: src/Cardano/AddressV3.cpp (hunk 5 above)
Comment (catenocrypt) [reply src/Cardano/AddressV3.cpp]: Suggestion worth considering, after checking usage I leave it like this though.

File: src/Cardano/AddressV3.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    Kind kind;
+
+    /// key1: spending key or account key
+    Data key1;
+
+    /// group key (in case of Group address, empty otherwise)
+    Data groupKey;
+
+    /// Used in case of legacy address (V2)
+    TW::Cardano::AddressV2* legacyAddressV2;
```
Comment (hewigovens) [inline src/Cardano/AddressV3.h]:
```suggestion
    AddressV2* legacyAddressV2;
```

Comment (catenocrypt) [reply src/Cardano/AddressV3.h]: done

**Comment by hewigovens** [comment]
let's resolve the conflict and merge it?

**Comment by catenocrypt** [comment]
> let's resolve the conflict and merge it?

done.

**Comment by Hazhir007** [comment]
Hi , I like to help create the sign of cardano tx , is it possible to give me or direct me to some sources on how to sign a cardano transaction? thanks

**Pull request merged**